File: tools/ioflow/ioflow-read.py
#!/usr/bin/python
# 
# ioflow-read.py        End-to-end IO Flow Tracer for read() syscall. 
# 
# Usage: 
#    ./ioflow-read.py           # Default time threshold: 1ms for syscalls and 0.2ms for requests. 
#    ./ioflow-read.py -t 5      # Print data if syscall latency exceeds 5ms or request latency exceeds 0.2ms.
#
#
# Copyright (c) Google LLC
# Licensed under the Apache License, Version 2.0 (the "License")
#
# 25-Aug-2020   Haoning Chen   Created this. 

from __future__ import print_function
from bcc import BPF 
import argparse

# parse arguments
examples = """examples:
    ./ioflow-read.py              # Trace read io flow. Default time threshold: 1ms for syscalls and 0.2ms for requests.
    ./ioflow-read.py -s 5         # Set syscall threshold to 5ms. Print syscall data if its latency exceeds 5 ms.
    ./ioflow-read.py -s 5 -r 0.5  # Set syscall threshold to 5ms and request threshold to 0.5 ms.
"""
parser = argparse.ArgumentParser(
    description="End-to-end IO Flow Tracer for read IOs",
    formatter_class=argparse.RawDescriptionHelpFormatter,
    epilog=examples)
parser.add_argument("-s", "--sys_thres", type=float, default=1.,
    help="Set syscall threshold in ms. Default to be 1.0")
parser.add_argument("-r", "--rq_thres", type=float, default=0.2,
    help="Set request threshold in ms. Default to be 0.2")
args = parser.parse_args()


# load Tracer program 
with open("ioflow-read.c") as src_f:
    bpf_text = src_f.read()
with open("ioflow-common.h") as comm_f:
    comm_text = comm_f.read()

# load BPF program
bpf = BPF(text=bpf_text.replace("[IMPORT_COMM]", comm_text, 1))

# file system layer:
bpf.attach_kprobe(event="vfs_read", fn_name="vfs_read_entry")
bpf.attach_kprobe(event="generic_file_read_iter", fn_name="page_cache_entry")
bpf.attach_kprobe(event="__do_page_cache_readahead", fn_name="read_page_entry")
bpf.attach_kprobe(event="ext4_mpage_readpages", fn_name="ext4_read_page_entry")
# block layer:
bpf.attach_kprobe(event="submit_bio", fn_name="block_entry")
bpf.attach_kretprobe(event="submit_bio", fn_name="block_return")
bpf.attach_kprobe(event="bio_split", fn_name="split_entry")
bpf.attach_kretprobe(event="bio_split", fn_name="split_return")
bpf.attach_kprobe(event="bio_attempt_front_merge", fn_name="merge_entry")
bpf.attach_kretprobe(event="bio_attempt_front_merge", fn_name="merge_return")
bpf.attach_kprobe(event="bio_attempt_back_merge", fn_name="merge_entry")
bpf.attach_kretprobe(event="bio_attempt_back_merge", fn_name="merge_return")
# async request handling:
bpf.attach_kprobe(event="blk_account_io_start", fn_name="rq_create")
if BPF.get_kprobe_functions(b'blk_start_request'):
	bpf.attach_kprobe(event="blk_start_request", fn_name="rq_issue")
bpf.attach_kprobe(event="blk_mq_start_request", fn_name="rq_issue")
bpf.attach_kprobe(event="blk_account_io_done", fn_name="rq_done")
# end of IO:
bpf.attach_kretprobe(event="vfs_read", fn_name="vfs_read_return")


# header
print("Tracing read I/Os. Time threshold: %.1f ms for syscalls and %.1f ms for requests. " % (args.sys_thres, args.rq_thres))
print("2 types of emit output with the following formats:\n")

print("[REQUEST] %6s %6s %9s %9s %11s %10s %14s %8s %6s\n" % 
    ("PID", "IO_NUM", "TOTAL_LAT", "CREATE_TS", "QUEUE_LAT", "SERV_LAT", "SECTOR", "LEN", "DISK"))

print("[SYSCALL] %6s %6s %9s %9s %11s %10s %14s %9s %8s %8s %5s %11s %9s %9s %5s %11s %9s %9s %5s %10s %8s %8s %6s\n" % 
    ("PID", "IO_NUM", "TOTAL_LAT", "VFS_LAT", "PGCACHE_LAT", "READPG_LAT", "EXT4READPG_LAT", "BLK_START", 
    "BLK_LAT", "BLK_END", "COUNT", "SPLIT_START", "SPLIT_LAT", "SPLIT_END", "COUNT", "MERGE_START", 
    "MERGE_LAT", "MERGE_END", "COUNT", "COMMAND", "OFFSET", "SIZE", "FILE"))

print("Hit Ctrl-C to end and display histograms.\n")


# process syscall events
def print_syscall_event(cpu, data, size):
    event = bpf["syscall_events"].event(data)
    total = float(event.total) / 1000
    if total >= (args.sys_thres * 1000):
        ts_blk_start = float(event.ts_blk_start - event.ts_vfs) / 1000 if event.ts_blk_start > 0 else 0.
        ts_blk_end = float(event.ts_blk_end - event.ts_vfs) / 1000 if event.ts_blk_end > 0 else 0.
        ts_split_start = float(event.ts_split_start - event.ts_vfs) / 1000 if event.ts_split_start > 0 else 0.
        ts_split_end = float(event.ts_split_end - event.ts_vfs) / 1000 if event.ts_split_end > 0 else 0.
        ts_merge_start = float(event.ts_merge_start - event.ts_vfs) / 1000 if event.ts_merge_start > 0 else 0.
        ts_merge_end = float(event.ts_merge_end - event.ts_vfs) / 1000 if event.ts_merge_end > 0 else 0.

        print("[SYSCALL] %6s %6s %9.3f %9.3f %11.3f %10.3f %14.3f %9.3f %8.3f %8.3f %5s %11.3f %9.3f %9.3f %5s %11.3f %9.3f %9.3f %5s %10s %8s %8s %6s" 
            % (event.pid, event.seq_num, total, float(event.vfs)/1000, float(event.pgcache)/1000, float(event.readpg)/1000, 
            float(event.ext4readpg)/1000, ts_blk_start, float(event.blk)/1000, ts_blk_end, event.cnt_blk, 
            ts_split_start, float(event.split)/1000, ts_split_end, event.cnt_split, ts_merge_start, float(event.merge)/1000, 
            ts_merge_end, event.cnt_merge, event.cmd_name, event.offset, event.size, event.file_name))


# process request events
def print_rq_event(cpu, data, size):
    event = bpf["rq_events"].event(data)
    total = float(event.queue + event.service) / 1000
    if total >= (args.rq_thres * 1000):
        print("[REQUEST] %6s %6s %9.3f %9.3f %11.3f %10.3f %14s %8s %6s" % 
            (event.pid, event.seq_num, total, float(event.ts_create)/1000, float(event.queue)/1000, 
            float(event.service)/1000, event.sector, event.len, event.disk_name))


# loop with callback to print_event
bpf["syscall_events"].open_perf_buffer(print_syscall_event, page_cnt=64)
bpf["rq_events"].open_perf_buffer(print_rq_event, page_cnt=128)
while True:
    try:
        bpf.perf_buffer_poll()
    except KeyboardInterrupt:
        break

print()
bpf["hist_vfs"].print_log2_hist("VFS (us)")
print()
bpf["hist_pgcache"].print_log2_hist("Page Cache Entry (us)")
print()
bpf["hist_readpg"].print_log2_hist("Read Page (us)")
print()
bpf["hist_ext4readpg"].print_log2_hist("EXT4 Read Page (us)")
print()
bpf["hist_blk"].print_log2_hist("Block Entry (us)")
print()
bpf["hist_split"].print_log2_hist("Bio Split (us)")
print()
bpf["hist_merge"].print_log2_hist("Bio Merge (us)")
print()
bpf["hist_rq_queue"].print_log2_hist("Request Queue (us)")
print()
bpf["hist_rq_service"].print_log2_hist("Request Service (us)")
exit()

File: tools/ioflow/ioflow-read.c
/* 
 * ioflow-read.c        C source code for read tracer. 
 *
 *
 * Copyright (c) Google LLC
 * Licensed under the Apache License, Version 2.0 (the "License")
 *
 * 25-Aug-2020   Haoning Chen   Created this.
 */

#include <uapi/linux/ptrace.h>
#include <linux/blkdev.h>
#include <linux/fs.h>

#define IO_FLOW_READ 1
// Use python string replacement to import common header code
[IMPORT_COMM]

BPF_HASH(syscall_map, u64, struct val_t);    // Map of syscall data, with pid as key
BPF_HASH(request_map, struct request *, struct rqval_t);    // Map of request info for asynchronous request handling

BPF_HISTOGRAM(hist_vfs);
BPF_HISTOGRAM(hist_pgcache);
BPF_HISTOGRAM(hist_readpg);
BPF_HISTOGRAM(hist_ext4readpg);
BPF_HISTOGRAM(hist_blk);
BPF_HISTOGRAM(hist_split);
BPF_HISTOGRAM(hist_merge);
// Async request handling histograms
BPF_HISTOGRAM(hist_rq_queue);
BPF_HISTOGRAM(hist_rq_service);

BPF_PERF_OUTPUT(syscall_events);
BPF_PERF_OUTPUT(rq_events);


int vfs_read_entry(struct pt_regs *ctx, struct file *file) {
    // Filter asyc and direct IO
    if (!(file->f_op->read_iter))
        return 0;
    
    u64 pid = bpf_get_current_pid_tgid();
    struct val_t val = {0};
    comm_vfs_entry(&val, file);
    syscall_map.update(&pid, &val);
    return 0;
}

int page_cache_entry(struct pt_regs *ctx) {
    u64 pid = bpf_get_current_pid_tgid();
    struct val_t *valp = syscall_map.lookup(&pid);
    if (valp && valp->ts_pgcache == 0)
        valp->ts_pgcache = bpf_ktime_get_ns();
    return 0;
}

int read_page_entry(struct pt_regs *ctx) {
    u64 pid = bpf_get_current_pid_tgid();
    struct val_t *valp = syscall_map.lookup(&pid);
    if (valp && valp->ts_readpg  == 0)
        valp->ts_readpg = bpf_ktime_get_ns();
    return 0;
}

int ext4_read_page_entry(struct pt_regs *ctx) {
    u64 pid = bpf_get_current_pid_tgid();
    struct val_t *valp = syscall_map.lookup(&pid);
    if (valp && valp->ts_ext4readpg  == 0)
        valp->ts_ext4readpg = bpf_ktime_get_ns();
    return 0;
}

/* BLock layer */  
int block_entry(struct pt_regs *ctx) {
    u64 ts = bpf_ktime_get_ns();
    u64 pid = bpf_get_current_pid_tgid();
    struct val_t *valp = syscall_map.lookup(&pid);
    if (valp) {
        comm_block_entry(valp, ts);
    }
    return 0;
}

int block_return(struct pt_regs *ctx) {
    u64 ts = bpf_ktime_get_ns();
    u64 pid = bpf_get_current_pid_tgid();
    struct val_t *valp = syscall_map.lookup(&pid);
    if (valp) {
        comm_block_return(valp, ts);
    }
    return 0;
}

int split_entry(struct pt_regs *ctx) {
    u64 ts = bpf_ktime_get_ns();
    u64 pid = bpf_get_current_pid_tgid();
    struct val_t *valp = syscall_map.lookup(&pid);
    if (valp) {
        comm_split_entry(valp, ts);
    }
    return 0;
}

int split_return(struct pt_regs *ctx) {
    u64 ts = bpf_ktime_get_ns();
    u64 pid = bpf_get_current_pid_tgid();
    struct val_t *valp = syscall_map.lookup(&pid);
    if (valp) {
        comm_split_return(valp, ts);
    }
    return 0;
}

int merge_entry(struct pt_regs *ctx) {
    u64 ts = bpf_ktime_get_ns();
    u64 pid = bpf_get_current_pid_tgid();
    struct val_t *valp = syscall_map.lookup(&pid);
    if (valp) {
        comm_merge_entry(valp, ts);
    }
    return 0;
}

int merge_return(struct pt_regs *ctx) {
    u64 ts = bpf_ktime_get_ns();
    u64 pid = bpf_get_current_pid_tgid();
    struct val_t *valp = syscall_map.lookup(&pid);
    if (valp) {
        comm_merge_return(valp, ts);
    }
    return 0;
}

// Async request handling
int rq_create(struct pt_regs *ctx, struct request *rq) {
    // Still in the syscall process's context now. 
    u64 ts = bpf_ktime_get_ns();
    u64 pid = bpf_get_current_pid_tgid();
    struct val_t *valp = syscall_map.lookup(&pid);
    if (valp) {
        struct rqval_t rqval = {0};
        comm_rq_create(&rqval, valp, rq, pid >> 32, ts);
        request_map.insert(&rq, &rqval);
    }
    return 0;
}

// The request is issued to device driver.
int rq_issue(struct pt_regs *ctx, struct request *rq) {
    // Async to the syscall process now. 
    u64 ts = bpf_ktime_get_ns();
    struct rqval_t *rqvalp = request_map.lookup(&rq);
    comm_rq_issue(rqvalp, ts);
    return 0;
}

// The request is done.
int rq_done(struct pt_regs *ctx, struct request *rq) {
    u64 ts = bpf_ktime_get_ns();
    struct rqval_t *rqvalp = request_map.lookup(&rq);
    if (rqvalp) {
        struct rqdata_t rqdata = {0};
        comm_rq_done(&rqdata, rqvalp, ts);
        rq_events.perf_submit(ctx, &rqdata, sizeof(rqdata));

        hist_rq_queue.increment(bpf_log2l(rqdata.queue / 1000));
        hist_rq_service.increment(bpf_log2l(rqdata.service / 1000));
        
        request_map.delete(&rq);
    }
    return 0;
}

// The end of the read IO
int vfs_read_return(struct pt_regs *ctx) {
    struct file *file = (struct file *) PT_REGS_PARM1(ctx);
    ssize_t size = (ssize_t) PT_REGS_RC(ctx);
    u64 pid = bpf_get_current_pid_tgid();

    struct val_t *valp = syscall_map.lookup(&pid);
    if (!valp) 
        return 0;
    if (valp->ts_pgcache == 0) {
        syscall_map.delete(&pid);
        return 0;
    }
    
    // Populate output struct 
    u64 ts_end = bpf_ktime_get_ns();
    struct data_t data = {0};
    comm_vfs_return(&data, valp, pid >> 32, ts_end, size);

    if (valp->ts_pgcache > valp->ts_vfs) 
        data.vfs = valp->ts_pgcache - valp->ts_vfs;
    else 
        // Direct IO. Directly falls to block layer. 
        data.vfs = data.ts_blk_start - valp->ts_vfs;
    
    if (valp->ts_readpg > valp->ts_pgcache)
        data.pgcache = valp->ts_readpg - valp->ts_pgcache;
    else if (data.ts_blk_start > valp->ts_pgcache)
        data.pgcache = data.ts_blk_start - valp->ts_pgcache;
    else 
        data.pgcache = ts_end - valp->ts_pgcache;

    if (valp->ts_readpg != 0 && valp->ts_ext4readpg > valp->ts_readpg)
        data.readpg = valp->ts_ext4readpg - valp->ts_readpg;

    if (valp->ts_ext4readpg != 0 && data.ts_blk_start > valp->ts_ext4readpg)
        data.ext4readpg = data.ts_blk_start - valp->ts_ext4readpg;
    
    syscall_events.perf_submit(ctx, &data, sizeof(data));
    syscall_map.delete(&pid);

    // update histograms
    if (data.vfs)
        hist_vfs.increment(bpf_log2l(data.vfs / 1000));
    if (data.pgcache)
        hist_pgcache.increment(bpf_log2l(data.pgcache / 1000));
    if (data.readpg)
        hist_readpg.increment(bpf_log2l(data.readpg / 1000));
    if (data.ext4readpg)
        hist_ext4readpg.increment(bpf_log2l(data.ext4readpg / 1000));
    if (data.blk)
        hist_blk.increment(bpf_log2l(data.blk / 1000));
    if (data.split)
        hist_split.increment(bpf_log2l(data.split / 1000));
    if (data.merge)
        hist_merge.increment(bpf_log2l(data.merge / 1000));
    
    return 0;
}
